common/http/platform/beast/http_resumer.cpp

```diff
@@ -0,0 +1,454 @@
+	if (!user_writer) {
+		auto err = MakeError(
+			DownloadResumerError, "The user did not set a BodyWriter to write the data into");
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		return;
+	}
+	client_.state_.user_body_writer = user_writer;
+}
+void HeaderHandlerFunctor::HandleNextResponse(ExpectedIncomingResponsePtr exp_resp) {
+	// Subsequent responses shall not call the user handler. If an error occurs, either
+	// schedule the next AsyncCall directly or save it and let the body handler forward
+	// it to the user
+	if (!exp_resp) {
+		client_.logger_.Warning(exp_resp.error().String());
+
+		auto err = ScheduleNextResumeRequest();
+		if (err != error::NoError) {
+			client_.logger_.Error(err.String());
+			client_.Cancel();
+			user_body_handler_(expected::unexpected(err));
+		}
+		return;
+	}
+	auto resp = exp_resp.value();
+
+	auto exp_content_range_header = resp->GetHeader("Content-Range");
+	if (!exp_content_range_header) {
+		client_.logger_.Error(exp_content_range_header.error().String());
+		client_.Cancel();
+		client_.state_.err = exp_content_range_header.error();
+		return;
+	}
+
+	auto exp_content_range = ParseRangeHeader(exp_content_range_header.value());
+	if (!exp_content_range) {
+		client_.logger_.Error(exp_content_range.error().String());
+		client_.Cancel();
+		client_.state_.err = exp_content_range.error();
+		return;
+	}
+
+	auto content_range = exp_content_range.value();
+	if (content_range.size != 0 && content_range.size != client_.state_.content_length) {
+		auto err = MakeError(
+			DownloadResumerError,
+			"Size of artifact changed after download was resumed (expected "
+				+ to_string(client_.state_.content_length) + ", got "
+				+ to_string(content_range.size) + ")");
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		client_.state_.err = err;
+		return;
+	}
+	if ((content_range.range_end != client_.state_.content_length - 1)
+		|| (content_range.range_start > client_.state_.offset)) {
+		auto err = MakeError(
+			DownloadResumerError,
+			"HTTP server did not return expected range. Expected "
+				+ to_string(client_.state_.offset) + "-"
+				+ to_string(client_.state_.content_length - 1) + ", got "
+				+ to_string(content_range.range_start) + "-" + to_string(content_range.range_end));
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		client_.state_.err = err;
+	} else if (content_range.range_start < client_.state_.offset) {
+		// The server resumed from an earlier offset, but we can still use the data
+		auto buffer = client_.state_.buffer.get();
+		auto overriden = client_.state_.offset - content_range.range_start;
+		client_.logger_.Warning(
+			"HTTP server returned a range with an earlier start. Expected "
+			+ to_string(client_.state_.offset) + ", got " + to_string(content_range.range_start)
+			+ ". The last " + to_string(overriden) + " bytes will be overriden");
+		buffer->erase(buffer->end() - overriden, buffer->end());
```

Comment:
What if `overridden` is larger than the buffer size? If this is difficult to handle, then I think producing error here is perfectly fine. There are limits to how far we should go in tolerating badly behaving servers.

Reply:
It should not be possible for it to be bigger (it would mean that we got a range start < 0 !) but you have a point that I should sanity check it and error accordingly. I'll do.

Reply:
> It should not be possible for it to be bigger (it would mean that we got a range start < 0 !)

Can't it? Maybe I'm misreading it, but to me it seems like `client_.state_.offset` can be *much* larger than buffer, and if the `range_start` is far behind, then the `overridden` result can easily be larger than buffer. But maybe I'm misunderstanding `offset`. Isn't it the full download offset? So it could be hundreds of megabytes.

Reply:
I rewrote quite a bit this part. Have a second look in the next round :eyes: